forgerock-core/src/main/java/org/forgerock/android/auth/biometric/BiometricAuth.kt

```diff
@@ -0,0 +1,245 @@
+     * @return the Activity that hosts the prompt.
+     */
+    val activity: FragmentActivity,
+
+    /**
+     * Return the Biometric Authentication listener used to return authentication result.
+     * @return the Biometric Authentication listener.
+     */
+    var biometricAuthListener: BiometricAuthCompletionHandler? = null,
+
+    /**
+     * The description to be displayed on the prompt.
+     * @return the description as String.
+     */
+    private val description: String? = null,
+
+    /**
+     * The type of authenticator to be displayed.
+     * @return the authenticatorType as Int.
+     */
+    var authenticatorType: AuthenticatorType = AuthenticatorType.WEAK
+) {
+    private var biometricManager: BiometricManager? = null
+
+    private var fingerprintManager: FingerprintManager? = null
+
+    /**
+     * Return the mechanism used to lock and unlock the device.
+     * @return the KeyguardManager instance.
+     */
+    var keyguardManager: KeyguardManager? = null
+        private set
+    private var promptInfo: BiometricPrompt.PromptInfo? = null
+
+
+    /*
+     * Starts authentication process.
+     */
+    fun authenticate() {
+        // if biometric only, try biometric prompt
+        if (!allowDeviceCredentials) {
+            if (hasBiometricCapability()) {
+                initBiometricAuthentication()
+            } else {
+                debug(
+                    TAG, "allowDeviceCredentials is set to false, but no biometric " +
+                            "hardware found or enrolled."
+                )
+                biometricAuthListener?.onError(
+                    ERROR_NO_BIOMETRICS, "It requires " +
+                            "biometric authentication. No biometric hardware found or enrolled."
+                )
+            }
+            return
+        }
+
+        // API 29 and above, use BiometricPrompt
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            initBiometricAuthentication()
+            return
+        }
+
+        // API 23 - 28, check enrollment with FingerprintManager once BiometricPrompt might not work
+        if(hasEnrolledWithFingerPrint()) {
+            initBiometricAuthentication()
+            return
+        }
+
+        // API 23 or higher, no biometric, fallback to device credentials
+        if (hasDeviceCredential()) {
+            initDeviceCredentialAuthentication()
+        } else {
+            debug(
+                TAG, "This device does not support required security features." +
+                        " No Biometric, device PIN, pattern, or password registered."
+            )
+            biometricAuthListener?.onError(
+                ERROR_NO_DEVICE_CREDENTIAL, "This device does " +
+                        "not support required security features. No Biometric, device PIN, pattern, " +
+                        "or password registered."
+            )
+        }
```

Comment:
I'd recommend "normalizing" all these conditions and reducing the first condition into a single level by abstracting the two nested conditions out. 

As of right now, each condition is quite unique and therefore has an increased cognitive overhead for the reader. In other words, I think this is sorta switch-statement-like, but isn't because of the uniqueness of how each case is considered. If you abstract out the uniqueness into functions, like you have some of these, it would normalize the conditions and ease in code understanding. I think :)

Reply:
i tried to keep it little bit more readable, removed the nested functions, this is a shared code , i can refactor even further when i attack device verifier story